### ISSUE TITLE
 Feature Request: Add abortEarly Mode to Stop Validation at First Error

### ISSUE THREAD

**Issue by mhdalmajid** [colinhacks/zod#3884]
Zod is a fantastic validation library, but many developers, myself included, face performance challenges when validating data with multiple rules, especially for expensive operations like database queries or HTTP calls. Currently, Zod validates all rules and surfaces as many errors as possible, which is excellent for debugging but not ideal for all use cases.

**Problem:**
- In some scenarios, such as form validation or validating HTTP payloads, it’s more efficient to stop validation as soon as the first error is encountered. For example:
  - Validating an email and password where only the first error per field is needed for the user.
  - Avoiding unnecessary, expensive operations (e.g., server requests) after detecting an initial failure.

**Proposed Solution:**
Introduce an optional `abortEarly` mode to Zod, similar to how Joi or Yup handle this. When enabled, Zod should stop validation at the first error and return a single issue. This could be implemented as an option in methods like `.parse()`, `.safeParse()`, and their async counterparts.

Example:

```javascript
schema.parse(data, { abortEarly: true });
schema.safeParse(data, { abortEarly: true });
await schema.parseAsync(data, { abortEarly: true });
await schema.safeParseAsync(data, { abortEarly: true });
```

**Benefits:**
- Improved performance by skipping unnecessary validations.
- Better control over error reporting in scenarios where only one error per field is needed.
- Enhanced flexibility for developers with diverse use cases.

**Additional Notes:**
This issue has been discussed extensively in [Issue #1403](https://github.com/colinhacks/zod/issues/1403), [Issue #1606](https://github.com/colinhacks/zod/issues/1606), and abort on signal [Issue #1915](https://github.com/colinhacks/zod/issues/1915).
 
Many developers have expressed a need for this feature, and alternative solutions like `superRefine` often result in less readable and more verbose code. 

For example:
- Form validation where only the first error message for each field is displayed.
- Avoiding redundant checks when basic rules like `required` fail.

Please reconsider adding this feature to improve developer experience and efficiency.

Thank you for considering this request!

**Comment by Rick-Phoenix** [comment]
This feature would be really good to have. 
Sometimes it's useful to insert some database operations in refine or superrefine because it creates a single source of errors which can be more easily handled and formatted along the way.

But when there is both input validation and database logic on the same zod validation chain, it would make a lot of sense to make zod stop at the first part (if inputs are not valid) before querying the database needlessly. 

However at the moment this is only doable if the inputs are validated with superRefine one by one and adding the fatal flag and early return to each one of them, which obviously defeats the purpose of it.

Ideally, there would be two methods to stop validation early:
- A single "abortEarly" flag, as suggested by OP and others. 
- A method like .checkPoint() where you can divide the validation chain in several "checkpoints" where you can handle the errors based on which portion of the validation chain they occurred.

**Comment by MentalGear** [comment]
Yes, it's still highly relevant, since as you yourself identified there is a huge community demand for it. Please reopen and prioritize.

**Comment by mhdalmajid** [comment]
still relevant 😀

**Comment by dewita97** [comment]
Hi guys,

I found something interesting in the library repository and in the documentation. 

Since May 2025, the abortEarly option has been commented out in the ts schema: [link](https://github.com/colinhacks/zod/blob/main/packages/zod/src/v4/core/schemas.ts)

Here is the related commit: [link](https://github.com/colinhacks/zod/commit/85928549a7c2aff851d8625588ed7aa2e394831f#diff-f84405ad925758ccb224a20fe7ca8c6ee375d7faa88431059e008077550f8cd9)

If abortEarly is present in the file, chances are we'll get that option one day!

**Comment by PierBover** [comment]
Hopefully this will be implemented. Having to use `superRefine()` just to be able to abort early can become unnecessarily tedious.

As an example I have a custom async validation that needs to check something with the API. All the previous validators need to pass before the HTTP request can be made. For reasons... I also need the validation to behave differently when submitting the form or while the user is writing.

Being able to model this using regular function chains or Zod mini would be much more straightforward.

**Comment by aaronfulkerson** [comment]
> Hopefully this will be implemented. Having to use `superRefine()` just to be able to abort early can become unnecessarily tedious.
> 
> As an example I have a custom async validation that needs to check something with the API. All the previous validators need to pass before the HTTP request can be made. For reasons... I also need the validation to behave differently when submitting the form or while the user is writing.
> 
> Being able to model this using regular function chains or Zod mini would be much more straightforward.

Aren't you supposed to be able to abort in .refine() also?

https://zod.dev/api?id=refine

I'm here because refine does _not_ seem to abort early as I had expected it to.

**Comment by mhdalmajid** [comment]
> > Hopefully this will be implemented. Having to use `superRefine()` just to be able to abort early can become unnecessarily tedious.
> > 
> > As an example I have a custom async validation that needs to check something with the API. All the previous validators need to pass before the HTTP request can be made. For reasons... I also need the validation to behave differently when submitting the form or while the user is writing.
> > 
> > Being able to model this using regular function chains or Zod mini would be much more straightforward.
> 
> Aren't you supposed to be able to abort in .refine() also?
> 
> https://zod.dev/api?id=refine
> 
> I'm here because refine does _not_ seem to abort early as I had expected it to.

it dose not abort , it continues validating the fields and return with errors

**Comment by MadL0rd** [comment]
This feature is still relevant

**Comment by MaximusHaximus** [comment]
Still hoping to get this Some Day ™️

**Comment by Schrubitteflau** [comment]
> Still hoping to get this Some Day ™️

The creator mentionned `{abort: true}` i other issues, I haven't tested it but it seems to go in the right direction: https://github.com/colinhacks/zod/issues/1606#issuecomment-3115444489 and https://github.com/colinhacks/zod/issues/3753#issuecomment-3116588724
